chore(Grid): add kb for selectall with onRead (#2697)
telerik/blazor-docs: knowledge-base/grid-select-all-onread.md

---

> title: Select All Items With CheckBox when Using OnRead

Reviewer: ```suggestion
title: Select All Grid Items With CheckBox when Using OnRead
```

> description: How Select All Items With CheckBox when Using OnRead in the Grid for Blazor

Reviewer: ```suggestion
description: How to select all Grid rows with checkBox when using the OnRead event in the Telerik Grid for Blazor.
```

type: how-to
page_title: How Select All Items With CheckBox when Using OnRead
slug: grid-kb-select-all-onread
position: 
tags: grid, selection, select all, checkbox, onread
ticketid: 1562945
res_type: kb
---

## Environment

<table>
    <tbody>
        <tr>
            <td>Product</td>
            <td>Grid for Blazor</td>
        </tr>
    </tbody>
</table>

## Description

I am binding the Grid through the `OnRead` event and I am using CheckBox selection with `SelectAllMode` set to `GridSelectAllMode.All`. However, when the user clicks the SelectAll CheckBox, only the items on the current page are selected, not all items in the dataset. How to ensure all items will be selected when clicking the SelectAll CheckBox.

## Solution

> The described behavior is expected as when using the `OnRead` event, the Grid operates only with the current set/page of data. The component does not have information for all the items in your datasource and [thus it cannot select them all](slug:grid-selection-row#selection-and-paging).

Reviewer: This paragraph belongs to a **Cause** section.

To ensure all item will be selected upon clicking the SelectAll CheckBox when using the `OnRead` event, you can implement a custom approach: 

1. Use [`HeaderTemplate` for the `CheckboxColumn`](slug:components/grid/columns/checkbox#header-template) and add a custom [CheckBox component](slug:checkbox-overview) so you can have full control over its behavior.
1. Handle the [`OnChange` event](slug:checkbox-events#onchange) of the CheckBox to track when the user checks/unchecks it to manage the selected items. 

>     - When the CheckBox is checked, request all the data from your datasource and assign it to the `SelectedItems` collection, so you have all items selected. You may want to cache all the data for further usage, so you don't need to request it additionally upon clicking the custom SelectAll CheckBox.

Reviewer: This may not be acceptable if the data item count is too large. Another option is to raise a flag, so that all new items received in `OnRead` (e.g. during paging) are added to `SelectedItems`.

                                     Indeterminate="@(SelectedEmployees.Count() >= 1 && SelectedEmployees.Count() < Total)"
                                     IndeterminateChanged="@HandleIndeterminateChanged"
                                     OnChange="@HandleSelectAll"/>                    
                }
            </HeaderTemplate>
        </GridCheckboxColumn>
        <GridColumn Field=@nameof(Employee.ID) />
        <GridColumn Field=@nameof(Employee.Name)/>
        <GridColumn Field=@nameof(Employee.HireDate) Title="Hire Date" />
    </GridColumns>
</TelerikGrid>

@code {
    private List<Employee> SourceData { get; set; }

    private IEnumerable<Employee> SelectedEmployees { get; set; } = Enumerable.Empty<Employee>();

    private bool SelectAllCheckBoxValue { get; set; }

    private int Total { get; set; } = 0;

    //Track when the user checks/unchecks the custom SelectAll CheckBox
    private void HandleSelectAll()
    {
        if (SelectAllCheckBoxValue)
        {
            // Request all the data and assign it to the SelectedItems collection. Here we use the already available SourceData for brevity.

>             SelectedEmployees = SourceData;

Reviewer: Requesting all the data may not be possible or desirable, otherwise the Grid won't be using `OnRead`. Consider something like this instead: https://blazorrepl.telerik.com/mTaROTvI23ir7Eqy34 (from 1680863)

        }
        else
        {
            SelectedEmployees = Enumerable.Empty<Employee>();
        }
    }    

    // Ensure that clicking the custom SelectAll CheckBox when it is in its indeterminate state will always select all items. 
    // This mimics the default SelectAll CheckBox behavior.
    private void HandleIndeterminateChanged(bool value)
    {
        if (!value)
        {
            SelectAllCheckBoxValue = true;
        }
    }

    protected async Task ReadItems(GridReadEventArgs args)
    {
        var datasourceResult = SourceData.ToDataSourceResult(args.Request);

        args.Data = datasourceResult.Data;
        args.Total = datasourceResult.Total;

        Total = datasourceResult.Total;
    }

    protected override void OnInitialized()
    {
        SourceData = GenerateData();
    }

    private List<Employee> GenerateData()
    {
        var result = new List<Employee>();

>         var rand = new Random();

Reviewer: Use the static `Random.Shared` instead.

---
title: Select All Grid Items With CheckBox when Using OnRead
description: How to select all Grid rows with checkBox when using the OnRead event in the Telerik Grid for Blazor.
type: how-to
page_title: How to Select All Grid Items With CheckBox when Using OnRead
slug: grid-kb-select-all-onread
position: 
tags: grid, selection, select all, checkbox, onread

> ticketid: 1562945

Reviewer: ```suggestion
ticketid: 1562945, 1680863
```
